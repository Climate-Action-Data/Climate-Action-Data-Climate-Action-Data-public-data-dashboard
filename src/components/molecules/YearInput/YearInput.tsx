import { ChangeEvent, FC, useEffect, useState } from 'react'
import { Input, InputGroup, Text, VStack } from '@chakra-ui/react'
import { DateFormats } from '@/@types/DateFormats'
import { YEAR_INPUT_REGEX } from '@/@types/Regex'

interface YearInputProp {
  label: string
  value: number | undefined
  maxYear?: number
  minYear?: number
  onChange: (year: number | undefined) => void
  onKeyDown: (event: React.KeyboardEvent<HTMLInputElement>) => void
}

const YearInput: FC<YearInputProp> = (prop) => {
  const { label, value, maxYear, minYear, onChange, onKeyDown } = prop

  const [textInputValue, setTextInputValue] = useState<string>(``)
  const [isInvalid, setIsInvalid] = useState<boolean>(false)

  useEffect(() => {
    setTextInputValue(value ? value?.toString() : ``)
  }, [value])

  const handleOnChange = (event: ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value
    if (value === ``) {
      setTextInputValue(value)
    }

    if (YEAR_INPUT_REGEX.test(value)) {
      setTextInputValue(value)
    }
  }

  const updateValue = () => {
    if (textInputValue === ``) {
      setIsInvalid(false)
      onChange(undefined)
    }

    if (!YEAR_INPUT_REGEX.exec(textInputValue)) {
      setIsInvalid(true)
    }

    const yearNumber = parseInt(textInputValue)
    if ((maxYear && yearNumber > maxYear) || (minYear && yearNumber < minYear)) {
      setIsInvalid(true)
    }

    setIsInvalid(false)
    onChange(yearNumber)
  }

  const handleOnBlur = () => {
    updateValue()
  }

  const handleOnKeyDown = (event: React.KeyboardEvent<HTMLInputElement>) => {
    if (event.key === `Enter`) {
      console.log(`===========   ${textInputValue}`)
      updateValue()
      onKeyDown(event)
    }
  }

  return (
    <VStack alignItems="start">
      <Text>{label}</Text>
      <InputGroup width="143px" alignItems="top">
        <Input
          variant="yearInput"
          placeholder={DateFormats.YYYY}
          borderBottomColor="lightGray.600"
          value={textInputValue}
          onChange={handleOnChange}
          onBlur={handleOnBlur}
          isInvalid={isInvalid}
          errorBorderColor="red"
          onKeyDown={handleOnKeyDown}
        />
      </InputGroup>
    </VStack>
  )
}

export default YearInput
