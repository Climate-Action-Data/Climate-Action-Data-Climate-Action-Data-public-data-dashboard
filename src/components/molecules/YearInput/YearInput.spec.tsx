import { render, screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import YearInput from '@/components/molecules/YearInput/YearInput'

describe(`YearInput`, () => {
  const tYear = 2023
  const tLabel = `YearInput`
  const user = userEvent.setup({ delay: null })

  afterAll(() => {
    jest.useRealTimers()
  })

  test(`renders a YearInput`, () => {
    const tOnChange = jest.fn()

    const { container } = render(<YearInput label={tLabel} value={tYear} onChange={tOnChange} />)

    expect(container).toMatchSnapshot()
  })

  test(`renders a YearInput with undefined value`, () => {
    const tOnChange = jest.fn()

    const { container } = render(<YearInput label={tLabel} value={undefined} onChange={tOnChange} />)

    expect(container).toMatchSnapshot()
  })

  test(`renders a YearInput and type in a correct value`, async () => {
    const tOnChange = jest.fn()

    const { container } = render(<YearInput label={tLabel} value={undefined} onChange={tOnChange} />)

    await user.type(screen.getByRole(`textbox`), `2023`)
    await user.tab()

    expect(tOnChange).toHaveBeenCalledTimes(1)
    expect(container).toMatchSnapshot()
  })

  test(`renders a YearInput and type in a wrong value`, async () => {
    const tOnChange = jest.fn()

    const { container } = render(<YearInput label={tLabel} value={undefined} onChange={tOnChange} />)

    await user.type(screen.getByRole(`textbox`), `wrong`)
    await user.tab()

    // eslint-disable-next-line no-magic-numbers
    expect(tOnChange).toHaveBeenCalledTimes(2)
    expect(container).toMatchSnapshot()
  })

  test(`renders a YearInput and type in a value earlier than minYear`, async () => {
    const tOnChange = jest.fn()

    const { container } = render(<YearInput label={tLabel} value={undefined} onChange={tOnChange} minYear={tYear} />)

    await user.type(screen.getByRole(`textbox`), `2022`)
    await user.tab()

    expect(container).toMatchSnapshot()
  })

  test(`renders a DateInput and type in a value later than maxDate`, async () => {
    const tOnChange = jest.fn()

    const { container } = render(<YearInput label={tLabel} value={undefined} onChange={tOnChange} maxYear={tYear} />)

    await user.type(screen.getByRole(`textbox`), `2024`)
    await user.tab()

    expect(container).toMatchSnapshot()
  })
})
