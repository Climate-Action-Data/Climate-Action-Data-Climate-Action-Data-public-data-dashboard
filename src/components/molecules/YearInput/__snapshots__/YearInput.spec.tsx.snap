// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YearInput renders a DateInput and type in a value later than maxDate 1`] = `
<div>
  <div
    class="chakra-stack css-m3jj3s"
  >
    <p
      class="chakra-text css-0"
    >
      YearInput
    </p>
    <div
      class="chakra-input__group css-1yb751z"
      data-group="true"
    >
      <input
        class="chakra-input css-p8rns9"
        placeholder="yyyy"
        value="2024"
      />
    </div>
  </div>
</div>
`;

exports[`YearInput renders a YearInput 1`] = `
<div>
  <div
    class="chakra-stack css-m3jj3s"
  >
    <p
      class="chakra-text css-0"
    >
      YearInput
    </p>
    <div
      class="chakra-input__group css-1yb751z"
      data-group="true"
    >
      <input
        class="chakra-input css-p8rns9"
        placeholder="yyyy"
        value="2023"
      />
    </div>
  </div>
</div>
`;

exports[`YearInput renders a YearInput and type in a correct value 1`] = `
<div>
  <div
    class="chakra-stack css-m3jj3s"
  >
    <p
      class="chakra-text css-0"
    >
      YearInput
    </p>
    <div
      class="chakra-input__group css-1yb751z"
      data-group="true"
    >
      <input
        class="chakra-input css-p8rns9"
        placeholder="yyyy"
        value="2023"
      />
    </div>
  </div>
</div>
`;

exports[`YearInput renders a YearInput and type in a value earlier than minYear 1`] = `
<div>
  <div
    class="chakra-stack css-m3jj3s"
  >
    <p
      class="chakra-text css-0"
    >
      YearInput
    </p>
    <div
      class="chakra-input__group css-1yb751z"
      data-group="true"
    >
      <input
        class="chakra-input css-p8rns9"
        placeholder="yyyy"
        value="2022"
      />
    </div>
  </div>
</div>
`;

exports[`YearInput renders a YearInput and type in a wrong value 1`] = `
<div>
  <div
    class="chakra-stack css-m3jj3s"
  >
    <p
      class="chakra-text css-0"
    >
      YearInput
    </p>
    <div
      class="chakra-input__group css-1yb751z"
      data-group="true"
    >
      <input
        class="chakra-input css-p8rns9"
        placeholder="yyyy"
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`YearInput renders a YearInput with undefined value 1`] = `
<div>
  <div
    class="chakra-stack css-m3jj3s"
  >
    <p
      class="chakra-text css-0"
    >
      YearInput
    </p>
    <div
      class="chakra-input__group css-1yb751z"
      data-group="true"
    >
      <input
        class="chakra-input css-p8rns9"
        placeholder="yyyy"
        value=""
      />
    </div>
  </div>
</div>
`;
